c/src/ml-api-service-training-offloading.c

```diff
@@ -0,0 +1,1058 @@
+    training_s)
+{
+  g_return_val_if_fail (training_s != NULL, ML_ERROR_INVALID_PARAMETER);
+
+  training_s->received_thread =
+      g_thread_new ("check_received_file", (GThreadFunc) check_received_data,
+      training_s);
+
+  g_mutex_lock (&training_s->received_lock);
+
+  while (!training_s->is_received) {
+    _ml_logd ("Wait to receive all data needed for model training");
+    g_cond_wait (&training_s->received_cond, &training_s->received_lock);
+    if (training_s->is_received == FALSE)
+      break;
+  }
+
+  g_mutex_unlock (&training_s->received_lock);
+  _ml_logd ("unlock, receive all data");
+
+  return training_s->is_received;
+}
+
+/**
+ * @brief replace path.
+ */
+static void
+_ml_service_training_offloading_replce_pipeline_data_path (ml_service_s * mls)
+{
+  guint changed = 0;
+  ml_training_services_s *training_s = NULL;
+
+  g_return_if_fail (mls != NULL);
+
+  training_s = (ml_training_services_s *)
+      ml_service_offloading_get_training_handle (mls);
+  g_return_if_fail (training_s != NULL);
+
+  if (training_s->type == ML_TRAINING_OFFLOADING_TYPE_SENDER) {
+    if (training_s->sender_pipe) {
+      training_s->sender_pipe =
+          _ml_replace_string (training_s->sender_pipe, APP_RW_PATH,
+          training_s->path, NULL, &changed);
+      _ml_logd ("@APP_RW_PATH@ is replaced, sender_pipe:%s",
+          training_s->sender_pipe);
+    }
+  } else {
+    if (training_s->receiver_pipe_json_str) {
+      training_s->trained_model_path =
+          _ml_replace_string (training_s->trained_model_path, APP_RW_PATH,
+          training_s->path, NULL, &changed);
+      training_s->receiver_pipe_json_str =
+          _ml_replace_string (training_s->receiver_pipe_json_str,
+          REMOTE_APP_RW_PATH, training_s->path, NULL, &changed);
+      training_s->receiver_pipe_json_str =
+          _ml_replace_string (training_s->receiver_pipe_json_str,
+          TRAINED_MODEL_FILE, training_s->trained_model_path, NULL, &changed);
+      _ml_logd
+          ("@REMOTE_APP_RW_PATH@ and @TRAINED_MODEL_FILE@ are replaced, receiver_pipe JSON string: %s",
+          training_s->receiver_pipe_json_str);
+    }
+  }
+}
+
+/**
+ * @brief Internal function to invoke ml-service event for new data.
+ */
+static int
+_invoke_event_new_data (ml_service_s * mls, const char *name,
+    const ml_tensors_data_h data)
+{
+  int status = ML_ERROR_NONE;
+  ml_service_event_cb_info_s cb_info = { 0 };
+  ml_information_h info = NULL;
+
+  g_return_val_if_fail (mls != NULL, ML_ERROR_INVALID_PARAMETER);
+  g_return_val_if_fail (data != NULL, ML_ERROR_INVALID_PARAMETER);
+  g_return_val_if_fail (name != NULL, ML_ERROR_INVALID_PARAMETER);
+
+  _ml_service_get_event_cb_info (mls, &cb_info);
+
+  if (cb_info.cb) {
+    /* Create information handle for ml-service event. */
+    status = _ml_information_create (&info);
+    if (status != ML_ERROR_NONE)
+      goto done;
+
+    if (name) {
```

Comment:
unnecessary condition (name is not null here, checked at line 500)

Reply:
Thank you.